File: src/main/java/net/jgp/books/spark/ch08/lab311_sql_joins/MySQLWithJoinToDatasetApp.java
package net.jgp.books.spark.ch08.lab311_sql_joins;

import java.util.Properties;

import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;

/**
 * MySQL injection to Spark, using the Sakila sample database.
 * 
 * @author jgp
 */
public class MySQLWithJoinToDatasetApp {

  /**
   * main() is your entry point to the application.
   * 
   * @param args
   */
  public static void main(String[] args) {
    MySQLWithJoinToDatasetApp app =
        new MySQLWithJoinToDatasetApp();
    app.start();
  }

  /**
   * The processing code.
   */
  private void start() {
    SparkSession spark = SparkSession.builder()
        .appName(
            "MySQL with join to Dataframe using JDBC")
        .master("local")
        .getOrCreate();

    // Using properties
    Properties props = new Properties();
    props.put("user", "root");
    props.put("password", "Spark<3Java");
    props.put("useSSL", "false");
    props.put("serverTimezone", "EST");

    // Builds the SQL query doing the join operation
    String sqlQuery =
        "select * "
            + "from actor, film_actor, film "
            + "where actor.actor_id = film_actor.actor_id "
            + "and film_actor.film_id = film.film_id";

    Dataset<Row> df = spark.read().jdbc(
        "jdbc:mysql://localhost:3306/sakila",
        "(" + sqlQuery + ") actor_film_alias",
        props);

    // Displays the dataframe and some of its metadata
    df.show(5);
    df.printSchema();
    System.out.println("The dataframe contains " + df
        .count() + " record(s).");
  }
}

File: src/main/python/lab310_sql_joins/mySQLWithJoinToDatasetApp.py
"""
 MySQL injection to Spark, using the Sakila sample database.

 @author rambabu.posa
"""
import logging
from pyspark.sql import SparkSession

# Creates a session on a local master
spark = SparkSession.builder.appName("MySQL with join to Dataframe using JDBC") \
    .master("local[*]").getOrCreate()

user = "root"
password = "Spark<3Java"
use_ssl="false"
mysql_url = "jdbc:mysql://localhost:3306/sakila?serverTimezone=EST"
dbtable = "actor"

df = spark.read.format("jdbc") \
    .option("url", mysql_url) \
    .option("user", user) \
    .option("dbtable", dbtable) \
    .option("password", password) \
    .load()


# Builds the SQL query doing the join operation
sqlQuery = """
    select actor.first_name, actor.last_name, film.title, 
    film.description 
    from actor, film_actor, film 
    where actor.actor_id = film_actor.actor_id 
    and film_actor.film_id = film.film_id
"""

df = df.select(sqlQuery)

# Displays the dataframe and some of its metadata
df.show(5)
df.printSchema()

logging.info("The dataframe contains {} record(s).".format(df.count()))

spark.stop()